### ISSUE TITLE
_SynchronizedBatchNorm's backward is redefined by its forward, right?

### ISSUE THREAD

**Issue by acgtyrant** [vacancy/Synchronized-BatchNorm-PyTorch#2]
I am reading the source codes of Synced BN in fyu/drn, zhanghang1989/PyTorch-SyncBatchNorm and your program at the same time.

I find that your program is different with the others, the others rewrite the BN function to pass the global mean and std to it while the official BN function calculate the local mean and std inside, but you just use simple expression like `output = (input - _unsqueeze_ft(mean)) * _unsqueeze_ft(inv_std * self.weight) + _unsqueeze_ft(self.bias)` in the `_SynchronizedBatchNorm` module which is inherited from the `torch.nn.modules.batchnorm`.

So I think this means that the `_SynchronizedBatchNorm`'s `backward` will be redefined automatically to be adjusted to the expression so that you do not need to rewrite the function like others program, right?

**Comment by vacancy** [comment]
Hi, @acgtyrant. Your understanding is correct. Actually, `nn.Module` does not explicitly declare any backward operation. It relies on `autograd.Function` for the back-propagation. In `_SynchronizedBatchNorm`, we have overwrite the `forward` method, and thus the gradient will be "implicitly" re-defined, as described in your comment.

The only reason for inheriting `torch.nn.modules.batchnorm` is that we re-use the `__init__` method and it also helps us check the compatibility 

N.B., we do not call the official fused batch-norm function because doing so will increase the communication complexity between multiple GPUs. (Currently we sync only `x.mean()` and `x.pow(2).mean()` not `x`).
